nogo fails using WORKSPACE.bazel after upgrade to 0.45.0

### What version of rules_go are you using?

0.45.0

### What version of gazelle are you using?

0.35.0

### What version of Bazel are you using?

```
bazel version
Build label: 7.0.0- (@non-git)
Build target: @@//src/main/java/com/google/devtools/build/lib/bazel:BazelServer
Build time: Tue Jan 1 00:00:00 1980 (315532800)
Build timestamp: 315532800
Build timestamp as int: 315532800
```

### Does this issue reproduce with the latest releases of all the above?

ya

### What operating system and processor architecture are you using?

Linux amd64

### What did you do?

Upgrade rules_go to 0.45.0 from 0.44.2.

Upon upgrading, nogo fails on things that are wrong and returns garbage output like:

```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
compilepkg: nogo: errors found by nogo during build-time code analysis:
bazel-out/k8-fastbuild/bin/path/to/file/1437484352.go:1:1: at least one file in a package should have a package comment (ST1000)
```

Note that only the file path was replaced, not the file name. Bazel literally returns a random number here.  Downgrading to 0.44.2 makes the issue go away.

### What did you expect to see?

no nogo failures


### What did you see instead?

gibberish  nogo failures

nvm, I repro'd it.

https://github.com/DolceTriade/monorepo/tree/minimal

```
INFO: Invocation ID: 8954761e-7090-4656-ae1e-a13279895b82
INFO: Analyzed 2 targets (0 packages loaded, 0 targets configured).
ERROR: /home/hmodi/monorepo/repro/subdir/mod/BUILD.bazel:10:8: GoCompilePkg repro/subdir/mod/mod_test_test.external.a failed: (Exit 1): builder failed: error executing GoCompilePkg command (from target //repro/subdir/mod:mod_test) bazel-out/k8-opt-exec-ST-13d3ddad9198/bin/external/go_sdk/builder_reset/builder compilepkg -sdk external/go_sdk -installsuffix linux_amd64 -src repro/subdir/mod/mod.go -src repro/subdir/mod/mod_test.go ... (remaining 26 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
compilepkg: nogo: errors found by nogo during build-time code analysis:
bazel-out/k8-fastbuild/bin/repro/subdir/mod/2404402859.go:1:1: at least one file in a package should have a package comment (ST1000)
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.143s, Critical Path: 0.04s
INFO: 2 processes: 2 internal.
ERROR: Build did NOT complete successfully
```

The file in that module:
```go
// Package mod does a bunch of stuff.
package mod

func Important() {
}
```

--- a/go/tools/builders/compilepkg.go
+++ b/go/tools/builders/compilepkg.go
@@ -272,7 +272,7 @@ func compileArchive(
 	compilingWithCgo := haveCgo && cgoEnabled
 
 	filterForNogo := func(slice []string) []string {
-		var filtered []string
+		filtered := make([]string, 0, len(slice))
 		for _, s := range slice {
 			// Do not subject the generated empty .go file to nogo checks.
 			if s != emptyGoFilePath {

--- a/go/tools/builders/compilepkg.go
+++ b/go/tools/builders/compilepkg.go
@@ -203,6 +203,7 @@ func compileArchive(
 	}
 	defer cleanup()
 
+	emptyGoFilePath := ""
 	if len(srcs.goSrcs) == 0 {
 		// We need to run the compiler to create a valid archive, even if there's nothing in it.
 		// Otherwise, GoPack will complain if we try to add assembly or cgo objects.
@@ -227,7 +228,7 @@ func compileArchive(
 			matched:  true,
 			pkg:      "empty",
 		})
-
+		emptyGoFilePath = emptyGoFile.Name()
 	}
 	packageName := srcs.goSrcs[0].pkg
 	var goSrcs, cgoSrcs []string
@@ -270,18 +271,24 @@ func compileArchive(
 	// constraints.
 	compilingWithCgo := haveCgo && cgoEnabled
 
+	filterForNogo := func(slice []string) []string {
+		var filtered []string
+		for _, s := range slice {
+			// Do not subject the generated empty .go file to nogo checks.
+			if s != emptyGoFilePath {
+				filtered = append(filtered, s)
+			}
+		}
+		return filtered
+	}
 	// When coverage is set, source files will be modified during instrumentation. We should only run static analysis
 	// over original source files and not the modified ones.
 	// goSrcsNogo and cgoSrcsNogo are copies of the original source files for nogo to run static analysis.
-	goSrcsNogo := goSrcs
-	cgoSrcsNogo := cgoSrcs
+	goSrcsNogo := filterForNogo(goSrcs)
+	cgoSrcsNogo := append([]string{}, cgoSrcs...)
 
 	// Instrument source files for coverage.
 	if coverMode != "" {
-		// deep copy original source files for nogo static analysis, avoid being modified by coverage.
-		goSrcsNogo = append([]string{}, goSrcs...)
-		cgoSrcsNogo = append([]string{}, cgoSrcs...)
-
 		relCoverPath := make(map[string]string)
 		for _, s := range coverSrcs {
 			relCoverPath[abs(s)] = s

--- a/tests/core/nogo/generate/empty_test.go
+++ b/tests/core/nogo/generate/empty_test.go
@@ -63,8 +63,6 @@ package noempty
 
 import (
 	"fmt"
-	"path/filepath"
-	"strings"
 
 	"golang.org/x/tools/go/analysis"
 )
@@ -77,12 +75,11 @@ var Analyzer = &analysis.Analyzer{
 
 func run(pass *analysis.Pass) (interface{}, error) {
 	for _, f := range pass.Files {
-		pos := pass.Fset.PositionFor(f.Pos(), false)
-
-		if strings.HasSuffix(pos.Filename, filepath.Join(".", "_empty.go")) {
+		// Fail on any package that isn't the "simple_test"'s package ('simple') or 'main'.
+		if f.Name.Name != "simple" && f.Name.Name != "main" {
 			pass.Report(analysis.Diagnostic{
 				Pos:     0,
-				Message: fmt.Sprintf("Detected generated source code from rules_go: %s", pos.Filename),
+				Message: fmt.Sprintf("Detected generated source code from rules_go: package %s", f.Name.Name),
 			})
 		}
 	}

